src/proxy/ssh/server.js

```diff
@@ -0,0 +1,683 @@
+                  // Don't immediately end the connection on error, try to recover
+                  if (
+                    err.message.includes('early EOF') ||
+                    err.message.includes('unexpected disconnect')
+                  ) {
+                    console.log(
+                      '[SSH] Detected early EOF or unexpected disconnect on client side, attempting to recover',
+                    );
+                    // Try to keep the connection alive
+                    if (remoteGitSsh.connected) {
+                      console.log('[SSH] Connection still active, continuing');
+                      // Don't end the connection, let it try to recover
+                      return;
+                    }
+                  }
+                  // If we can't recover, then end the connection
+                  remoteGitSsh.end();
+                });
+
+                // Handle connection end
+                remoteGitSsh.on('end', () => {
+                  console.log('[SSH] Remote connection ended');
+                });
+
+                // Handle connection close
+                remoteGitSsh.on('close', () => {
+                  console.log('[SSH] Remote connection closed');
+                });
+
+                // Add a timeout to ensure the connection is closed if it hangs
+                const connectionTimeout = setTimeout(() => {
+                  console.log('[SSH] Connection timeout, ending connection');
+                  remoteGitSsh.end();
+                }, 300000); // 5 minutes timeout for large repositories
+
+                // Clear the timeout when the connection is closed
+                remoteGitSsh.on('close', () => {
+                  clearTimeout(connectionTimeout);
+                });
+              },
+            );
+          });
+
+          remoteGitSsh.on('error', (err) => {
+            console.error('[SSH] Remote SSH error:', err);
+
+            // If authentication failed and we're using the client key, try with the proxy key
+            if (
+              err.message.includes('All configured authentication methods failed') &&
+              clientKey &&
+              connectionOptions.privateKey !==
+                require('fs').readFileSync(config.getSSHConfig().hostKey.privateKeyPath)
+            ) {
+              console.log('[SSH] Authentication failed with client key, trying with proxy key');
+
+              // Create a new connection with the proxy key
+              const proxyGitSsh = new Client();
+
+              // Set up connection options with proxy key
+              const proxyConnectionOptions = {
+                ...connectionOptions,
+                privateKey: require('fs').readFileSync(
+                  config.getSSHConfig().hostKey.privateKeyPath,
+                ),
+                // Ensure these settings are explicitly set for the proxy connection
+                windowSize: 1024 * 1024, // 1MB window size
```

Comment:
Might be best to extract these (and the repeated ones from earlier in the file) into constants so they can be reused.